java/src/main/java/ai/rapids/cudf/MultiBufferDataSource.java

```diff
@@ -0,0 +1,225 @@
+    hostBuffers = new HostMemoryBuffer[numBuffers];
+    startOffsets = new long[numBuffers];
+
+    long currentOffset = 0;
+    for (int i = 0; i < numBuffers; i++) {
+      HostMemoryBuffer hmb = buffers[i];
+      hmb.incRefCount();
+      hostBuffers[i] = hmb;
+      startOffsets[i] = currentOffset;
+      currentOffset += hmb.getLength();
+    }
+    sizeInBytes = currentOffset;
+    this.allocator = allocator;
+  }
+
+  @Override
+  public long size() {
+    return sizeInBytes;
+  }
+
+  private int getStartBufferIndexForOffset(long offset) {
+    // It is super common to read from the start or end of a file (the header or footer)
+    // so special case them
+    if (offset == 0) {
+      return 0;
+    }
+    int startIndex = 0;
+    int endIndex = startOffsets.length - 1;
+    if (offset >= startOffsets[endIndex]) {
+      return endIndex;
+    }
+    while (startIndex != endIndex) {
+      int midIndex = (int)(((long)startIndex + endIndex) / 2);
+      long midStartOffset = startOffsets[midIndex];
+      if (offset >= midStartOffset) {
+        // It is either in mid or after mid.
+        if (midIndex == endIndex || offset <= startOffsets[midIndex + 1]) {
+          // We found it in mid
+          return midIndex;
+        } else {
+          // It is after mid
+          startIndex = midIndex + 1;
+        }
+      } else {
+        // It is before mid
+        endIndex = midIndex - 1;
+      }
+    }
+    return startIndex;
+  }
+
+
+  interface DoCopy<T extends MemoryBuffer> {
+    void copyFromHostBuffer(T dest, long destOffset, HostMemoryBuffer src,
+                            long srcOffset, long srcAmount);
+  }
+
+  private <T extends MemoryBuffer> long read(long offset, T dest, DoCopy<T> doCopy) {
+    long realOffset = Math.min(offset, sizeInBytes);
+    long realAmount = Math.min(sizeInBytes - realOffset, dest.getLength());
+
+    int index = getStartBufferIndexForOffset(realOffset);
+
+    HostMemoryBuffer buffer = hostBuffers[index];
+    long bufferOffset = realOffset - startOffsets[index];
+    long bufferAmount = Math.min(buffer.length - bufferOffset, realAmount);
+    long remainingAmount = realAmount;
+    long currentOffset = realOffset;
+    long outputOffset = 0;
+
+    while (remainingAmount > 0) {
+      doCopy.copyFromHostBuffer(dest, outputOffset, buffer,
+          bufferOffset, bufferAmount);
+      remainingAmount -= bufferAmount;
+      outputOffset += bufferAmount;
+      currentOffset += bufferAmount;
+      index++;
+      if (index < hostBuffers.length) {
+        buffer = hostBuffers[index];
+        bufferOffset = currentOffset - startOffsets[index];
+        bufferAmount = Math.min(buffer.length - bufferOffset, remainingAmount);
+      }
+    }
+
+    return realAmount;
+  }
+
+  @Override
+  public HostMemoryBuffer hostRead(long offset, long amount) {
+    long realOffset = Math.min(offset, sizeInBytes);
+    long realAmount = Math.min(sizeInBytes - realOffset, amount);
```

Comment:
Negative offset or amount will do some bad things here.  Since this is public, seems prudent to check here rather than assume the entry point will always be via `DataSource::hostReadBuff`.

Reply:
This comment has not been addressed.

Reply:
It was addressed in a different place by https://github.com/rapidsai/cudf/pull/14254/commits/8e66e63a3eb1d2ff8fc9c941c7d23beea14d5f74

I thought it was more of a generic problem so I fixed it in a more generic place.

Reply:
I saw that commit, but it doesn't address the issue here.  This function is public and therefore could be called directly, rather than only from the JNI.  The fix in 8e66e63 only works if it's called via hostReadBuff, and even then that fix only addresses negative amounts.  I believe negative offsets are still problematic even in the JNI-called path.

Reply:
If `offset` or `amount` is negative we will get a `realAmount` that is either too large or negative and/or a `realOffset` that is negative. In all of those cases we are going through a HostMemoryBuffer API that has asserts on it to protect against going off the limits if assertions are enabled and if they are not you are out of luck. This code is not going to violate that. But just to make debugging simpler if ti does happen I will add in asserts here too.